core/trie/proof.go

```diff
@@ -0,0 +1,80 @@
+package trie
+
+import (
+	"errors"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type Membership int
+
+const (
+	Member Membership = iota
+	NonMember
+)
+
+type ProofNode struct {
+	LeftHash  *felt.Felt
+	RightHash *felt.Felt
+}
+
+// https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs#L514
+func GetProof(leaf *felt.Felt, tri *Trie) ([]ProofNode, error) {
+	leafKey := tri.feltToKey(leaf)
+	nodesToLeaf, err := tri.nodesFromRoot(&leafKey)
+	if err != nil {
+		return nil, err
+	}
+	proofNodes := make([]ProofNode, len(nodesToLeaf)-1)
+
+	getHash := func(tri *Trie, key *Key) (*felt.Felt, error) {
+		keyFelt := key.Felt()
+		node, err := tri.GetNode(&keyFelt)
+		if err != nil {
+			return nil, err
+		}
+		return node.Hash(key, crypto.Pedersen), nil
+	}
+
+	for i, sNode := range nodesToLeaf[:len(nodesToLeaf)-1] {
+		leftHash, err := getHash(tri, sNode.node.Left)
+		if err != nil {
+			return nil, err
+		}
+		rightHash, err := getHash(tri, sNode.node.Right)
+		if err != nil {
+			return nil, err
+		}
+		proofNodes[i] = ProofNode{
+			LeftHash:  leftHash,
+			RightHash: rightHash,
+		}
+	}
+	return proofNodes, nil
+}
+
+// verifyProof checks if `leafPath` leads from `root` to `leafHash` along the `proofNodes`
+// https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs#L2006
+func VerifyProof(root *felt.Felt, leafPath *Key, leafHash felt.Felt, proofNodes []ProofNode, hashFunc hashFunc) (Membership, error) {
```

Comment:
Since the presence of an error affects Membership (i.e., if err is nil, then Membership is Member, and vice versa), I propose simplifying the signature:

Either leave only the error (a nil error indicates that the proof is correct, i.e., Member)
Or return a boolean as the first return value.
Personally, I'm in favor of the first option.